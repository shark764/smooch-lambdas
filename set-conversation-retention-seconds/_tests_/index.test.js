jest.mock('aws-sdk');
jest.mock('smooch-core/lib/smooch');

global.process.env = {
  AWS_REGION: 'us-east-1',
  ENVIRONMENT: 'dev',
  DOMAIN: 'domain',
  smooch_api_url: 'mock-amooch-api-url',
};

const mockGetSecretValue = jest
  .fn(() => {})
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        id: '5e31c81640a22c000f5d7c55',
        secret: 'secret',
      }),
    }),
  }));

const mockScanPromise = jest.fn().mockImplementation(() => ({
  Items: [
    { id: '5e31c81640a22c000f5d7f28', type: 'app' },
    { id: '5e31c81640a22c000f5d7f29', type: '' },
  ],
}));
const mockScan = jest.fn().mockImplementation(() => ({
  promise: mockScanPromise,
}));

const event = {};

jest.mock('aws-sdk', () => ({
  config: {
    update: jest.fn(),
  },
  SecretsManager: jest.fn().mockImplementation(() => ({
    getSecretValue: mockGetSecretValue,
  })),
  SQS: jest.fn().mockImplementation(() => ({
    getQueueUrl: jest
      .fn()
      .mockImplementation(() => ({ promise: jest.fn().mockImplementation(() => ({ QueueUrl: 'url://testurl' })) })),
    sendMessage: jest.fn().mockImplementation(() => ({ promise: jest.fn() })),
  })),
  DynamoDB: {
    DocumentClient: jest.fn().mockImplementation(() => ({
      scan: mockScan,
    })),
  },
}));

const mockSmoochUpdate = jest.fn().mockImplementation(() => ({
  promise: () => ({}),
}));

const mockSmoochCore = jest.fn(() => ({
  apps: {
    update: mockSmoochUpdate,
  },
}));

jest.mock('smooch-core', () => mockSmoochCore);

const { handler } = require('../index');

describe('set-conversation-retention-seconds', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  describe('Everything is successful', () => {
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('returns when the code runs without any error', async () => {
      const result = handler(event);
      expect(result).toBeTruthy();
    });

    describe('Walkthrough', () => {
      beforeEach(async () => {
        await handler(event);
      });

      it('passes in the correct arguments to secretClient.getSecretValue() to get cx credentials', async () => {
        expect(mockGetSecretValue.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to docClient.scan()', async () => {
        expect(mockScan.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to SmoochCore', async () => {
        expect(mockSmoochCore.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to smooch.apps.update()', async () => {
        expect(mockSmoochUpdate.mock.calls).toMatchSnapshot();
      });
    });
  });

  it('throws an error when there is a error scanning apps in DynamoDB', async () => {
    mockScanPromise.mockRejectedValueOnce(new Error('An Error has occurred trying to fetch apps in DynamoDB'));
    await expect(handler(event)).rejects.toMatchSnapshot();
  });

  it('throws an error when there is a problem updating Smooch appUser', async () => {
    mockSmoochUpdate.mockRejectedValueOnce(new Error());
    await expect(handler(event)).rejects.toMatchSnapshot();
  });
});
