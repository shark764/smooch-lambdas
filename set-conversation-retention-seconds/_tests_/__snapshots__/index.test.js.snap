// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`set-conversation-retention-seconds Everything is successful Walkthrough passes in the correct arguments to SmoochCore 1`] = `
Array [
  Array [
    Object {
      "keyId": "5e31c81640a22c000f5d7c55",
      "scope": "account",
      "secret": "secret",
      "serviceUrl": "mock-smooch-api-url",
    },
  ],
]
`;

exports[`set-conversation-retention-seconds Everything is successful Walkthrough passes in the correct arguments to docClient.scan() 1`] = `
Array [
  Array [
    Object {
      "TableName": "us-east-1-dev-smooch",
    },
  ],
]
`;

exports[`set-conversation-retention-seconds Everything is successful Walkthrough passes in the correct arguments to secretClient.getSecretValue() to get cx credentials 1`] = `
Array [
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-account",
    },
  ],
]
`;

exports[`set-conversation-retention-seconds Everything is successful Walkthrough passes in the correct arguments to smooch.apps.update() 1`] = `
Array [
  Array [
    "5e31c81640a22c000f5d7f28",
    Object {
      "settings": Object {
        "conversationRetentionSeconds": 172800,
      },
    },
  ],
]
`;

exports[`set-conversation-retention-seconds throws an error when there is a error scanning apps in DynamoDB 1`] = `[Error: An Error has occurred trying to fetch apps in DynamoDB]`;

exports[`set-conversation-retention-seconds throws an error when there is a problem updating Smooch appUser 1`] = `[Error: At least one of the apps was unable to update. See logs for details.]`;
